feat: update model to add type restrictions (#233)
openfga/openfga.dev: docs/content/concepts.mdx

</summary>

> A **user** is is a combination of a [type](#what-is-a-type) and an identifier.

Reviewer: ```suggestion
A **user** is is a combination of a [type](#what-is-a-type), an identifier and an optional relation.
```

For example,

- any identifier: e.g. `user:anne` or `user:4179af14-f0c0-4930-88fd-5570c7bf6f59`
- any object: e.g. `workspace:fb83c013-3060-41f4-9590-d3233a67938f`, `repository:auth0/express-jwt` or `organization:org_ajUc9kJ`
- a group or a set of users (also called a **userset**): e.g. `organization:org_ajUc9kJ#members`, which represents the set of users related to the object `organization:org_ajUc9kJ` as `member`

> - everyone, using the special syntax: `user:*`

Reviewer: ```suggestion
- everyone, using the special syntax: `*`
```

For more information, please see [Direct Access](./modeling/direct-access.mdx).

</details>

<details>
<summary>

## What Is A Directly Related User Type?

A **directly related user type** is an array specified in the type definition to indicate what types of users can be directly related to that relation.

</summary>

For the following model, only [relationship tuple](#what-is-a-relationship-tuple) with [user](#what-is-a-user) [type](#what-is-a-type) `user` may be assigned to document.

<AuthzModelSnippetViewer
  configuration={{
    type_definitions: [
      {
        type: 'document',
        relations: {
          viewer: {
            this: {},
          },
        },
        metadata: {
          relations: {
            viewer: { directly_related_user_types: [{ type: 'user' }] },
          },
        },
      },
      {
        type: 'user',
      },
    ],
  }}
/>

> Relationship tuple with user `user:anne` or `user:3f7768e0-4fa7-4e93-8417-4da68ce1846c` may be written for objects with type `document` and relation `viewer`. Tuples with user `workspace:auth0` or `folder:planning#editor` will be rejected. This will affect only relations that are [directly related](#what-are-direct-and-implied-relationships) and have [the direct relationship keyword ("this")](./configuration-language.mdx#the-direct-relationship-keyword) in their relation definition.

Reviewer: ```suggestion
Relationship tuple with user `user:anne` or `user:3f7768e0-4fa7-4e93-8417-4da68ce1846c` may be written for objects with type `document` and relation `viewer`, so writing `{"user": "user:anne","relation":"viewer","object":"document:roadmap"}` will succeed.
Relationship tuple writes with user type that is not allowed for the `viewer` relation on objects of type `document`, for example `workspace:auth0` or `folder:planning#editor` will be rejected, so writing `{"user": "folder:product","relation":"viewer","object":"document:roadmap"}` will fail.
This will affect only relations that are [directly related](#what-are-direct-and-implied-relationships) and have [the direct relationship keyword ("this")](./configuration-language.mdx#the-direct-relationship-keyword) in their relation definition.
```

  - <RelationshipTuplesViewer
      relationshipTuples={[
        {
          _description: 'Everyone (`*`) of type user is directly related to the document',

>           user: 'user:*',

Reviewer: ```suggestion
          user: '*',
```

</summary>

For example, in a case where you would like to indicate a certain document `document:new-roadmap` is publicly writable (i.e. has everyone of type user as an editor), you can add the following [relationship tuple](#what-is-a-relationship-tuple):

<RelationshipTuplesViewer
  relationshipTuples={[
    {

>       user: 'user:*',

Reviewer: ```suggestion
      user: '*',
```
